proto/regen/ecocredit/v1alpha1/tx.proto

```diff
@@ -0,0 +1,167 @@
+message MsgCreateClassRequest {
+
+    // designer is the address of the account which designed the credit class. The designer has special permissions
+    // to change the list of issuers and perform other administrative operations.
+    string designer = 1;
+
+    // issuers are the addresses of the approved issuers.
+    repeated string issuers = 2;
+
+    // metadata is any arbitrary metadata to attached to the credit class.
+    bytes metadata = 3;
+}
+
+// MsgCreateClassResponse is the Msg/CreateClass response type.
+message MsgCreateClassResponse {
+
+    // class_id is the unique ID of the newly created credit class.
+    string class_id = 1;
+}
+
+// MsgCreateBatchRequest is the Msg/CreateBatch request type.
+message MsgCreateBatchRequest {
+
+    // issuer is the address of the batch issuer.
+    string issuer = 1;
+
+    // class_id is the unique ID of the class.
+    string class_id = 2;
+
+    // issuance are the credits issued in the batch.
+    repeated BatchIssuance issuance = 3;
+
+    // metadata is any arbitrary metadata to attached to the credit batch.
+    bytes metadata = 4;
+
+    // BatchIssuance represents the issuance of some credits in a batch to a single recipient.
+    message BatchIssuance {
+
+        // recipient is the account of the recipient.
+        string recipient = 1;
+
+        // tradable_units are the units of credits in this issuance that can be traded by this recipient.
+        // Decimal values are acceptable.
+        string tradable_units = 2;
+
+        // retired_units are the units of credits in this issuance that are effectively retired by the issuer on receipt.
+        // Decimal values are acceptable.
+        string retired_units = 3;
+    }
+}
+
+// MsgCreateBatchResponse is the Msg/CreateBatch response type.
+message MsgCreateBatchResponse {
+
+    // batch_denom is the unique ID of the newly created batch.
+    string batch_denom = 1;
+}
+
+// MsgSendRequest is the Msg/Send request type.
+message MsgSendRequest {
+
+    // sender is the address of the account sending credits.
+    string sender = 1;
+
+    // sender is the address of the account receiving credits.
+    string recipient = 2;
+
+    // credits are the credits being sent.
+    repeated SendUnits credits = 3;
+
+    // SendUnits are the tradable and retired units of a credit batch to send.
+    message SendUnits {
+
+        // batch_denom is the unique ID of the credit batch.
+        string batch_denom = 1;
+
+        // tradable_units are the units of credits in this issuance that can be traded by this recipient.
+        // Decimal values are acceptable within the precision returned by Query/Precision.
+        string tradeable_units = 2;
+
+        // retired_units are the units of credits in this issuance that are effectively retired by the issuer on receipt.
+        // Decimal values are acceptable within the precision returned by Query/Precision.
+        string retired_units = 3;
```

Comment:
Seems strange to allow sending of pre-retired credits. IMO retiring should be a separate action done intentionally by the holder of credits.

We already cover the "retire on issuance" use case in the `BatchIssuance` message, so why is this needed here as well?

Reply:
Not necessarily. I believe a broker may specifically only sell credits that are retired on sale. We need to discuss with Ron It may or may not be a use case but better to be on the safe side

Reply:
agreed cf. https://github.com/regen-network/regen-ledger/pull/125#discussion_r517845025